### ISSUE TITLE
Handmade Math Speed

### ISSUE THREAD

**Issue by StrangeZak** [HandmadeMath/Handmade-Math#73]
One of the things ive been thinking about recently is how handmade-math's speed fairs up against the other major math libraries. 

Is there some sort of tool we can use to speed test all of our function just to get an idea where our bottlenecks are.

**Comment by bvisness** [comment]
I'm sure you could pretty easily write a benchmark program that performs common graphics applications and times the results.

**Comment by StrangeZak** [comment]
Yeah i was just kinda wondering if there was something like that already written for us. Because i dont mind writing our own, but just sort of a thought

**Comment by bvisness** [comment]
I spent some time tonight making a small benchmark program. I didn't get around to testing it against GLM or other libraries, but I did compare a couple operations with SSE enabled vs. SSE disabled.

```
Adding 100,000,000 random vectors - average time
With SSE:    7.090625 seconds
Without SSE: 7.328125 seconds
Speedup:     1.0335

Transposing 100,000,000 random matrices - average time
With SSE:    11.262500 seconds
Without SSE: 13.531250 seconds
Speedup:     1.2014
```

I can keep working on it soon.

**Comment by StrangeZak** [comment]
Oh wow nice! im just kinda interested to see how much room we have to optimize. Also how we fair against other libraries will be nice to see! Im not very good  at SSE so its nice to see the few things i did are faster

**Comment by RandyGaul** [comment]
You might have to look at the disassembly to make sure the scalar version wasn't vectorized for you by the compiler. Often times when I was looking at SSE perf stuff it was pretty difficult to put together good comparisons.

Personally I would love to see scalar vs SSE implementation of an algorithm, like maybe a cloth solver or something. Then test out an implementation with many load/stores. I heard load/store isn't a big deal anymore since 64bit will compile scalar ops with vector registers, making load/stores basically free. This contrasts a lot of older articles on SIMD implementations that say load/stores are expensive.

**Comment by bvisness** [comment]
My curiosity got the better of me and I tested everything against GLM, and tested both with and without compiler optimizations. Turns out Handmade Math is overall better, although the gains diminish a lot with optimization.

https://docs.google.com/spreadsheets/d/1e-jQt7789JEE-_UoQFmKPHB1Whcx5xTYROCo7snhe5s/edit?usp=sharing

However, a lot of the timing in these results is the overhead of generating all the random numbers for the test cases. I think I may jump in again sometime soon with a proper profiling tool and see how the actual functions fare.

**Comment by bvisness** [comment]
@RandyGaul I would love to dig into the disassembly at some point. I ran my latest batch of tests both with and without the `-O3` flag for gcc, so hopefully that can provide a better point of comparison, although it's still pretty rough.

**Comment by StrangeZak** [comment]
I feel like the reason the gains are SO small on optimized builds (-03) is because the optimizer will attempt to SIMD the code for you. Sadly we can not depend on the compiler to do it for us though, as each compiler will SIMD the code in a different way so it is easier to just flat out write it our self so we can prevent the optimizer from doing bad things. 

Also the test was just adding vec4's there's not really a easy fast or slow way to do it. Its just one instruction either way through SSE(_mm_add_ps).

**Comment by bvisness** [comment]
@StrangeZak There are three tabs on the spreadsheet! Take another look.

**Comment by StrangeZak** [comment]
Oh Wow! Our unoptimized SSE builds are VERY fast! That makes me pretty happy. 

The data i think that is really interesting here is the Martix Multiplication data. Id love to see the disassembly to see what we could do to squeeze 2 seconds or more out of our optimized SSE version.   Id love to beat the optimizer here, and im fairly sure we could.

In every other case optimized (with SSE enabled, which should always be enabled in my opinion) we are beating GLM which is really awesome!
